Topics/Development_Process/Build_Requirements/Introduction_to_make_and_Makefiles.md

```diff
@@ -0,0 +1,228 @@
+
+<pre>
+<i>VARIABLENAME</i> := <i>value, ...</i>
+</pre>
+
+Doing the same variable assignments above with simply expanded variables:
+```
+GCC := cc
+CC := $(GCC)
+GCC := gcc
+```
+
+Then, ```CC``` would evaluate to ``cc``, as determined at the time of assignment.  
+This behaviour makes recursively expanded variables usually more suited for static variables, while simply expanded variables are usually more suited for dynamic variables.  
+
+Additionally, to append to a variable, use ```+=```. For the following example, ```CC``` would evaluate to ```gcc -o```: 
+
+```
+CC := gcc
+CC += -o
+```
+
+#### Automatic Variables
+
+When you cannot manually reference a target or prerequisite, you can use automatic variables that are computed for each executed rule. These are some simple useful ones:
+- ```$@```: The target name
+- ```$<```: The first prerequisite's name
+- ```$^```: All prerequisite names
+- ```$?```: All prerequisite names newer than the target
+
+#### Pattern Rules:
+
+Instead of manually writing a target for each object file like in the previous simple Makefile, pattern rules can be used. They are used as a list of targets that the rule will apply to, along with a target pattern that extracts the stem out of the target name, and uses it to match to the prerequisite pattern. They are written in the following template:
+
+<pre>
+<i>targets ...</i> <i>target-pattern</i>: <i>prerequisite-patterns ...</i>
+      <i>recipe</i>
+      <i>...</i>
+</pre>
+
+For example, in the following example, the ```all``` rule would run the corresponding pattern rule for ```objects```, and for each of ```main.o``` and ```helper.o``` in ```objects```, the target pattern would strip out the respective stems ```main``` and ```helper``` and apply it to the prerequisite pattern, resulting in the corresponding prerequisites ```main.c``` and ```helper.c```. Then each rule would be run, and their compile recipe would be executed, producing ```main.o``` and ```helper.o```:
+
+```
+OBJECTS = main.o helper.o
+
+all: $(OBJECTS)
+
+$(OBJECTS): %.o: %.c
+        $(CC) -c $< -o $@
+```
+
+#### Using Functions
+
+Functions have a multitude of uses in Makefiles. Some simple useful ones are:
+- ```$(wildcard pattern)```: returns a a list of filenames that match the given pattern
+- ```$(foreach var,list,text)```: sets ```var``` to the evaluated ```text``` for each word in ```list```
+- ```$(if condition,true-part,false-part)```: returns ```true-part``` if ```condition``` is true, otherwise returns ```false-part```
+
+#### Additional Tips
+
+You can have ```make``` silently execute commands using the ```@``` symbol. For example the following:
+
+```
+all:
+	echo Hello world
+	@echo Hello world
+```
+
+Would produce:
+
+```bash
+$ make all
+>>> echo Hello world
+>>> Hello world
+>>> Hello world
+```
+
```

Comment:
Could you tie all of the above parts together? It’s not super clear how all of the individual parts would work together.

Reply:
That makes sense, will do.